File: AAAA_TOP200/三数之和.cpp


#include<stdio.h>
#include<bits/stdc++.h>
#include<iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());

        vector<vector<int>> res;
        int n =nums.size();

        for(int i=0;i<n;i++) {

            // 两个数组相等。
            if(i>0&&nums[i]==nums[i-1]) {
                continue;
            }

            // 确定第三个元素
            int c = n-1;

            int target = -nums[i];

            for(int b=i+1;b<n;b++) {
                if(b>i+1&&nums[b]==nums[b-1]) {
                    continue;
                }

                // 后面的递减
                while(b<c&&nums[b]+nums[c]>target) {
        			c--;
        		}

                if(b==c) break;//若两个指针相等了还无解，就不用再比下去了
                if(nums[b]+nums[c]==target) {
                	res.push_back({nums[i],nums[b],nums[c]});
            	}
            }
        }

        return res;
    }
};

int main() {

    vector<int> nums = {-1,0,1,2,-1,-4};
    vector<vector<int>> res;

    Solution s2;
    res = s2.threeSum(nums);
    for(auto i:res) {
        for(auto j:i) {
            cout<<j<<" ";
        }
        cout<<endl;
    }
    return 0;
}

File: AAAA_TOP200/三数之和.java
package com.AAAA.www;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {

    public static List<List<Integer>> threeSum(int[] nums) {

        List<List<Integer>> res = new ArrayList<>();

        int len = nums.length;
        if(nums==null||nums.length<3) {
            return res;
        }
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++) {
            if(nums[i]>0) {
                break;
            }

            if(i>0&&nums[i]==nums[i-1]) {
                continue;
            }

            // 最左和最右

            int L = i+1;
            int R = len-1;

            while (L<R) {
                int sum = nums[i] + nums[L] + nums[R];
                if(sum==0) {
                    res.add(Arrays.asList(nums[i],nums[L],nums[R]));
                    while (L<R) {
                        if(nums[L]==nums[L+1]) {
                            L++;
                        }
                    }
                    while (L<R) {
                        if(nums[R]==nums[R-1]) {
                            R--;
                        }
                    }
                    L++;
                    R--;
                }
                else if(sum<0) {
                    L++;
                }
                else if(sum>0) {
                    R--;
                }
            }
        }
        return res;

    }

    public static void main(String[] args) {
        int[] nums = {-1,0,1,2,-1,-4};
        List<List<Integer>> res;
        res = threeSum(nums);
        for(var i:res) {
            for(var j:i) {
                System.out.print(j+" ");
            }
            System.out.println();
        }
    }
}
